File: frontend/src/views/workbench/wodaFields.ts
import type { RecognitionFieldOption } from './waybillFieldCatalog'
import type { WodaExtractedItem } from './wodaStructure'

export type CustomWodaFieldCode =
  | 'custom_product_text'
  | 'custom_sales_attr1_text'
  | 'custom_quantity_text'
  | 'custom_sales_attr2_text'
  | 'custom_item_remark_text'

export type CustomWodaFieldMapping = {
  field_key: CustomWodaFieldCode
  field_label: string
  source: string
  source_field_key?: CustomWodaFieldCode
  line_index: number
  segment_text?: string
  extractor: string
  usage?: string
}

export type WodaItemTextField =
  | 'productText'
  | 'salesAttr1Text'
  | 'quantityText'
  | 'salesAttr2Text'
  | 'remarkText'

export const CUSTOM_WODA_FIELD_ORDER: CustomWodaFieldCode[] = [
  'custom_product_text',
  'custom_sales_attr1_text',
  'custom_quantity_text',
  'custom_sales_attr2_text',
  'custom_item_remark_text',
]

export const CUSTOM_WODA_FIELD_LABELS: Record<CustomWodaFieldCode, string> = {
  custom_product_text: '商品文字',
  custom_sales_attr1_text: '销售属性1',
  custom_quantity_text: '数量文字',
  custom_sales_attr2_text: '销售属性2',
  custom_item_remark_text: '备注字段',
}

export const WODA_FIELD_LABELS: Record<string, string> = {
  ...CUSTOM_WODA_FIELD_LABELS,
  custom_spec_text: '销售属性1',
  custom_size_text: '销售属性2',
}

export const SAMPLE_ROW_FIELD_BY_CODE: Record<CustomWodaFieldCode, WodaItemTextField> = {
  custom_product_text: 'productText',
  custom_sales_attr1_text: 'salesAttr1Text',
  custom_quantity_text: 'quantityText',
  custom_sales_attr2_text: 'salesAttr2Text',
  custom_item_remark_text: 'remarkText',
}

export const WODA_SKU_AUTO_FIELD_CODES = new Set([
  'custom_sales_attr1_text',
  'custom_sales_attr2_text',
  'custom_spec_text',
  'custom_size_text',
  'custom_item_remark_text',
])

export function isCustomWodaFieldCode(value: unknown): value is CustomWodaFieldCode {
  return CUSTOM_WODA_FIELD_ORDER.includes(value as CustomWodaFieldCode)
}

export function extractorForWodaFieldCode(code: CustomWodaFieldCode): string {
  if (code === 'custom_product_text') return 'product_text'
  if (code === 'custom_sales_attr1_text') return 'sales_attr1_text'
  if (code === 'custom_sales_attr2_text') return 'sales_attr2_text'
  if (code === 'custom_quantity_text') return 'quantity_text'
  if (code === 'custom_item_remark_text') return 'remark_text'
  return 'text'
}

export function wodaItemFieldText(item: WodaExtractedItem, code: string): string {
  if (code === 'custom_product_text') return item.productText || ''
  if (code === 'custom_sales_attr1_text' || code === 'custom_spec_text') {
    return item.salesAttr1Text || item.specText || ''
  }
  if (code === 'custom_sales_attr2_text' || code === 'custom_size_text') {
    return item.salesAttr2Text || item.sizeText || ''
  }
  if (code === 'custom_quantity_text') return item.quantityText || ''
  if (code === 'custom_item_remark_text') return item.remarkText || ''
  return ''
}

export function wodaRecognitionFieldOptions(): RecognitionFieldOption[] {
  return [
    {
      code: 'custom_product_text',
      label: CUSTOM_WODA_FIELD_LABELS.custom_product_text,
      meaning: '',
      recommended: true,
    },
    {
      code: 'custom_sales_attr1_text',
      label: CUSTOM_WODA_FIELD_LABELS.custom_sales_attr1_text,
      meaning: '',
      recommended: true,
    },
    {
      code: 'custom_sales_attr2_text',
      label: CUSTOM_WODA_FIELD_LABELS.custom_sales_attr2_text,
      meaning: '',
    },
    {
      code: 'custom_item_remark_text',
      label: CUSTOM_WODA_FIELD_LABELS.custom_item_remark_text,
      meaning: '勾选后参与商品主类识别，也作为 SKU 兜底字段',
    },
  ]
}

File: backend/app/services/woda_fields.py
from __future__ import annotations

from typing import Any


WODA_CUSTOM_FIELD_ITEM_KEYS = {
    "custom_product_text": "product_text",
    "custom_sales_attr1_text": "sales_attr1_text",
    "custom_sales_attr2_text": "sales_attr2_text",
    "custom_quantity_text": "quantity_text",
    "custom_item_remark_text": "remark_text",
}

WODA_EXTRACTOR_FIELD_CODES = {
    "product_text": "custom_product_text",
    "sales_attr1_text": "custom_sales_attr1_text",
    "spec_text": "custom_sales_attr1_text",
    "sales_attr2_text": "custom_sales_attr2_text",
    "size_text": "custom_sales_attr2_text",
    "quantity_text": "custom_quantity_text",
    "remark_text": "custom_item_remark_text",
}

WODA_SKU_SOURCE_FIELD_SCORES = {
    "custom_sales_attr1_text": 300,
    "custom_spec_text": 300,
    "spec_text": 260,
    "custom_sales_attr2_text": 200,
    "custom_size_text": 200,
    "product_short_text": 240,
    "product_full_text": 220,
    "custom_item_remark_text": 100,
    "buyer_remark": 120,
    "seller_remark": 110,
}

WODA_SKU_FALLBACK_FIELD_CODES = {"custom_item_remark_text", "buyer_remark", "seller_remark"}
WODA_PRODUCT_TEXT_FIELD_CODES = {"custom_product_text", "product_short_text", "product_full_text"}
DEFAULT_WODA_SKU_AUTO_FIELDS = ("custom_sales_attr1_text", "custom_sales_attr2_text")


def _text(value: Any) -> str:
    if value is None:
        return ""
    return str(value).strip()


def woda_mapping_source_field(mapping: Any, fallback: str) -> str:
    if not isinstance(mapping, dict):
        return fallback
    source_field = _text(mapping.get("source_field_key"))
    if source_field in WODA_CUSTOM_FIELD_ITEM_KEYS:
        return source_field
    extractor = _text(mapping.get("extractor"))
    return WODA_EXTRACTOR_FIELD_CODES.get(extractor, fallback)
